### ISSUE TITLE
SocketGroup lost on daemon-reload with unit moving away temporarily

### ISSUE THREAD

**Issue by martinpitt** [systemd/systemd#3171]
### Submission type
- [X] Bug report
- [ ] Request for enhancement (RFE)
### systemd version the issue has been seen with

229, master as of yesterday
### Used distribution

Debian unstable

In https://launchpad.net/bugs/1577001 we noticed that the Unix socket files created from a `.socket` unit that has a `SocketGroup=lxd` option sometimes change to group `root` on package install/upgrade. During that, the unit file is temporarily not present, and this happens when `daemon-reload` is called in between (which happens fairly often in package maintainer scripts). During that time, the `.socket` unit stays `active (running)` all the time.

This can be demonstrated with a series of shell commands: http://people.canonical.com/~pitti/tmp/systemd-socketgroup.sh is a runnable shell script for that which can be run in containers, this is a copy&paste variant:

```
U=/run/systemd/system/foo.socket

# reset state after previous runs of this script
systemctl stop foo.socket
rm -f $U /tmp/foo.sock
systemctl daemon-reload

cat <<EOF > $U
[Unit]
Description=foo socket
[Socket]
ListenStream=/tmp/foo.sock
SocketGroup=adm
SocketMode=0660
EOF

cat <<EOF > /run/systemd/system/foo.service
[Unit]
Description=foo service
[Service]
ExecStart=/bin/true
EOF
echo '--- initial start ---'
systemctl start foo.socket
systemctl show -p SocketGroup foo.socket
ls -l /tmp/foo.sock

echo '--- move away unit ---'
mv $U ${U}.disabled
systemctl daemon-reload
ls -l /tmp/foo.sock
systemctl show -p SocketGroup foo.socket

echo '--- put back unit ---'
mv ${U}.disabled $U
systemctl daemon-reload
ls -l /tmp/foo.sock
systemctl show -p SocketGroup foo.socket
```

This outputs:

```
--- initial start ---
SocketGroup=adm
srw-rw---- 1 root adm 0 May  2 11:41 /tmp/foo.sock
--- move away unit ---
srw-rw---- 1 root adm 0 May  2 11:41 /tmp/foo.sock
SocketGroup=
--- put back unit ---
srw-rw---- 1 root root 0 May  2 11:41 /tmp/foo.sock
SocketGroup=adm
```

As you see, after putting back the unit file and daemon-reloading, the socket becomes `root:root`. Arguably this is a bug in lxd's maintainer scripts as removing the package ought to stop `lxd.socket` first (and we'll fix that), but (1) there have been reports that this also happens on a clean package install (and I just haven't found a simpler way to reproduce this), and (2) this is still inconsistent behaviour: Either the socket should not be touched at all when the unit file is gone, or the complete information about the running unit should stay in systemd's brain and correctly serialized on daemon-reload.

**Comment by martinpitt** [comment]
I was hoping that this was just due to a wrong serialization of socket units:

```
--- a/src/core/socket.c
+++ b/src/core/socket.c
@@ -620,8 +620,8 @@ static void socket_dump(Unit *u, FILE *f, const char *prefix) {

         if (!isempty(s->user) || !isempty(s->group))
                 fprintf(f,
-                        "%sOwnerUser: %s\n"
-                        "%sOwnerGroup: %s\n",
+                        "%sSocketUser: %s\n"
+                        "%sSocketGroup: %s\n",
                         prefix, strna(s->user),
                         prefix, strna(s->group));

```

As `Owner{User,Group}` are not a valid directive and don't occur anywhere else in the code. However, this does not fix _this_ issue. I still wonder if this should be fixed anyway, as this might break something else? Or am I missing something here and `Owner*` is actually correct?

**Comment by poettering** [comment]
socket_dump() only outputs the human readable output that is shown by "systemd-analyze dump".

the label "socket-proxy" refers to systemd-socket-proxy btw, which I see no relation to here.

**Comment by poettering** [comment]
hmm, so we only invoke chown() on the sockets when we go throught the "starting" phase of the socket unit, which a reload should never trigger...

**Comment by martinpitt** [comment]
Right, I already assured that `socket_start()` and `socket_chown()` are not called at `daemon-reload` time. But `socket_coldplug()` is called, which calls `socket_open_fds()`, which seems to (re)create the socket with wrong permissions. I'm still debugging this further (just a bit slow as I'm on a company sprint this week).

**Comment by poettering** [comment]
well, the fds to the sockets should be passed across the reload, hence we shouldn't be opening any new sockets...

**Comment by evverx** [comment]
I'm running:

``` sh
bash-4.3# strace -o /pid-1 -tt -f -v -s 150 -p1 &

bash-4.3# /test-socket-group.sh

bash-4.3# fg
^C

bash-4.3# grep -F /tmp/foo.sock /pid-1
1     13:17:24.089942 read(13, "[Unit]\nDescription=foo socket\n[Socket]\nListenStream=/tmp/foo.sock\nSocketGroup=adm\nSocketMode=0660\n", 4096) = 98
1     13:17:24.137502 bind(18, {sa_family=AF_LOCAL, sun_path="/tmp/foo.sock"}, 16) = 0
35    13:17:24.142053 chown("/tmp/foo.sock", -1, 4) = 0
1     13:17:25.159028 read(13, "[Unit]\nDescription=foo socket\n[Socket]\nListenStream=/tmp/foo.sock\nSocketGroup=adm\nSocketMode=0660\n", 4096) = 98
1     13:17:25.161452 bind(11, {sa_family=AF_LOCAL, sun_path="/tmp/foo.sock"}, 16) = -1 EADDRINUSE (Address already in use)
1     13:17:25.161472 unlink("/tmp/foo.sock") = 0
1     13:17:25.161491 bind(11, {sa_family=AF_LOCAL, sun_path="/tmp/foo.sock"}, 16) = 0
```

So, we `unlink("/tmp/foo.sock")` somewhere.

**Comment by evverx** [comment]
Seems like [`socket_address_listen`](https://github.com/systemd/systemd/blob/825546ef761c829b7a22ccdb33981d3d4d23c2b9/src/basic/socket-label.c#L114):

``` c
                RUN_WITH_UMASK(~socket_mode) {
                        r = mac_selinux_bind(fd, &a->sockaddr.sa, a->size);
                        if (r == -EADDRINUSE) {
                                /* Unlink and try again */
                                unlink(a->sockaddr.un.sun_path);
                                if (bind(fd, &a->sockaddr.sa, a->size) < 0)
                                        return -errno;
                        } else if (r < 0)
                                return r;
                }
```

**Comment by martinpitt** [comment]
This is a very naïve patch which replaces the `socket_open_fds()` in `socket_coldplug()` with `socket_enter_start_chown()` (which calls `socket_open_fds()` and then chowns asynchronously):

```
--- a/src/core/socket.c
+++ b/src/core/socket.c
@@ -1466,6 +1466,8 @@ static void socket_set_state(Socket *s, SocketState state) {
         unit_notify(UNIT(s), state_translation_table[old_state], state_translation_table[state], true);
 }

+static void socket_enter_start_chown(Socket *s);
+
 static int socket_coldplug(Unit *u) {
         Socket *s = SOCKET(u);
         int r;
@@ -1507,9 +1509,7 @@ static int socket_coldplug(Unit *u) {
                    SOCKET_STOP_PRE_SIGTERM,
                    SOCKET_STOP_PRE_SIGKILL)) {

-                r = socket_open_fds(s);
-                if (r < 0)
-                        return r;
+                socket_enter_start_chown(s);
         }

         if (s->deserialized_state == SOCKET_LISTENING) {
```

This doesn't do proper error handling (`socket_enter_start_chown` returns void) and thus isn't a final patch, but it seems to work.

**Comment by poettering** [comment]
Hmm, so here's what happens:

first, the socket is properly up and everything. Then, you move the unit file away, and reload. For the reload we serialize the state to disk, and attach the opened fds to it. After the reload we deserialize state and try to match the opened fds to the new configuration. FDs which cannot be matched up to the new configuration are closed. Since you moved away the unit file this is the case here, the opened socket will simply be closed, leaving the unix socket node on disk around orphaned. Then, after moving the unit file into place again, and reloading, there's suddenly new configuration, for which no open socket exists. Hence a new socket is created, owned initially by the root user. And this ownership doesn't change anymore. This is because normally, when a socket is established, the socket unit transitions through the special "chown" state, where the the unix socket is chown()ed in a forked off process (the reason we do this in a forked off process instead of PID 1 itself, is that we need to resolve the user/group name, and that might involve talking to local or remote services – think ldap, nscd, ssssd, …  – and we can't do that synchronously from PID 1, since that could mean PID 1 is on the other end of the service too, and is supposed to activate the actual service, but can't since we still try to resolve the user synchronously → deadlock). Now, while reloading state we won't change a unit's state, the idea should really be after all, that reloads change configuration, not state. Hence we won't transition throught the "chown" state, and hence the socket won't be chowned to the right user/group...

Now, the question is what to do about this. Maybe the following:

a) First of all, when after a daemon reload a socket unit is found to be active, but doesn't actually have any fd attached to it anymore (which happens in the first part of the story here, because the configuration vanished), then we probably should notice, and automatically transition the socket unit to "stopped" state, with an explanation why in the logs. In your case above, this would mean the unit would end up being stopped after your reload.

b) after a daemon reload we should probably not try to open new sockets anymore, but instead print a warning that to some configured fds of a socket unit are not in established, and to establish them the user needs to restart the socket unit.

I hope that makes some sense.

**Comment by martinpitt** [comment]
Automatically stopping sockets that have no unit any more on daemon-reload sounds reasonable. But IMHO then we should also start them again if the unit comes back, to be symmetric. This seems a bit hairy to me, though.

So b) sounds simpler and safer to me; is the state of "unit disappeared from disk" represented in the `Unit` object somewhere? (I think `UnitLoadState UNIT_NOT_FOUND`?) Could/should we bypass `socket_coldplug()` for those, and do the warning instead?

**Comment by poettering** [comment]
> Automatically stopping sockets that have no unit any more on daemon-reload sounds reasonable. But IMHO then we should also start them again if the unit comes back, to be symmetric. This seems a bit hairy to me, though.

Well, a stopped unit is a stopped unit, we don't keep history about whether on a previous reload it was started. And I am pretty sure if we'd automatically restart units that 10 reloads earlier were running but then were automatically stopped for a couple of reloads, because their unitfiles vanished, this would also be very confusing to the user.

**Comment by poettering** [comment]
> So b) sounds simpler and safer to me; is the state of "unit disappeared from disk" represented in the Unit object somewhere? (I think UnitLoadState UNIT_NOT_FOUND?) Could/should we bypass socket_coldplug() for those, and do the warning instead?

Oh, it wasn't a) XOR b), I wanted a) AND b)... ;-)

I wouldn't really bind any of this to the explicit state of "has no unit file config on disk", but more to "we just reloaded a running socket unit, and unit file lists more ListenStream=/ListenDatagram= stanzas than we actually have fds for. Complain, but proceed, do not open any new fds, until the user explicitly invokes restart on the socket unit"

**Comment by poettering** [comment]
Fix waiting in #3202.

I decided to not put the socket into dead state when it lost all fds due to a daemon-reload. Instead we'll simply warn in this case too, and suggest people restart the socket unit to get the full socket settings into place again.

**Comment by poettering** [comment]
(I decided not to place the socket into a dead state, simply because – as mentioned above – a daemon reload should not result in a state change of units. And there's actually nothing inherently wrong with permitting a socket unit to stay urnning even if it has no open fds. Now that we warn about this, this should be OK really)
